connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoIngestionUtils.scala

```diff
@@ -70,53 +72,127 @@ object KustoIngestionUtils {
-    There are 2 options:
-    either to not have a mapping or create an explicit identity mapping. Since GenerateCSVMapping is requested explicitly
-    creating an identity mapping made the most appropriate fit
-     */
-    val sourceSchemaColumnTypes =
-      if (tableCreationMode == SinkTableCreationMode.CreateIfNotExist) {
-        sourceSchema.fields
-          .map(field => (field.name, getSparkTypeToKustoTypeMap(field.dataType)))
-          .toMap
-      } else {
-        Map.empty[String, String]
-      }
-    val notFoundSourceColumns =
-      sourceSchemaColumns.filter(c => !targetSchemaColumns.contains(c._1)).keys.toSet
-    if (notFoundSourceColumns.nonEmpty && targetSchema != null && targetSchema.nonEmpty) {
-      // TODO Add
-      if (includeSourceTransforms) {
+  }
+
+  private def extractSourceSchemaColumns(sourceSchema: StructType): Map[String, Int] = {
+    sourceSchema.fields.zipWithIndex.map(c => (c._1.name, c._2)).toMap
+  }
+
+  private def extractSourceSchemaTypes(
+      sourceSchema: StructType,
+      tableCreationMode: SinkTableCreationMode): Map[String, String] = {
+    if (tableCreationMode == SinkTableCreationMode.CreateIfNotExist) {
+      sourceSchema.fields
+        .map(field => (field.name, getSparkTypeToKustoTypeMap(field.dataType)))
+        .toMap
+    } else {
+      Map.empty[String, String]
+    }
+  }
+
+  private def findMissingSourceColumns(
+      sourceSchemaColumns: Map[String, Int],
+      targetSchemaColumns: Map[String, String]): Set[String] = {
+    sourceSchemaColumns.filter(c => !targetSchemaColumns.contains(c._1)).keys.toSet
+  }
+
+  private def validateSchemaCompatibility(
+      writeMode: WriteMode,
+      notFoundSourceColumns: Set[String],
+      targetSchemaColumns: Map[String, String],
+      includeSourceLocationTransform: Boolean): Unit = {
+
+    if (includeSourceLocationTransform && !targetSchemaColumns.contains(
+        KustoConstants.SourceLocationColumnName)) {
+      throw SchemaMatchException(
+        "addSourceLocationTransform is set to true, but the target schema does not contain the " +
+          s"column '${KustoConstants.SourceLocationColumnName}'. " +
+          s"Please ensure that the target table has this column defined.")
+    }
+
+    if (notFoundSourceColumns.nonEmpty && targetSchemaColumns != null && targetSchemaColumns.nonEmpty) {
+      // The drift between source and target schema is allowed for queued ingestion only if includeSourceLocationTransform is true
+      if (writeMode == WriteMode.Queued && includeSourceLocationTransform) {
         KustoDataSourceUtils.logWarn(
           this.getClass.getSimpleName,
           s"Source schema has columns that are not present in the target: ${notFoundSourceColumns.mkString(",")}. " +
-            s"However, since the option 'addSourceLocationTransform' is set to true, the ingestion will continue.")
+            s"Since the option 'addSourceLocationTransform' is set to true, the ingestion will continue.")
```

Comment:
Lets explain our logic here 
"In order to preserve behavior of "no-mapping like" behavior for the client where we do not fail on schema drift as no validation is made"